### PR TITLE
decimal: specialized fast paths for Decimal64/128 operations

### PR DESCRIPTION
**Pull request by aunjgr**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [x] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #24097

## What this PR does / why we need it:
Optimizations for TPC-H Q1 SF100: 15.5s → 8.2s (47% speedup).

- Decimal64 SUM: branchless Add128Unchecked, count-based null tracking
- Decimal128 SUM: identity path, count-based null tracking
- SUM/AVG batch slice access: pre-extract vals[] and nulls before loop
- Decimal64→Decimal128 cast: direct slice write with branchless sign extension
- Decimal128 multiply: inline fast path for D64-origin values (single bits.Mul64)
- Decimal128 add/sub: bits.Add64/Sub64 for all cases (same-scale and different-scale)
- Decimal64 add/sub: same-scale/diff-scale split with pre-scaled constants
- Decimal64 multiply: branchless D64×D64→D128 batch kernel via framework
- Coercion fix: D64 ± int stays D64 (eliminates 3 full-column D64→D128 casts in Q1)
- Diff-scale optimization: pre-scale constant once, scale-into-rs for vectors
- Generalized decimalBatchArith[TIn, TOut] supporting mixed input/output types
- Consistent d{width}{Op} naming (d64Mul, d128Add, etc.)
- Remove dead decimal128AddArray/SubArray/MultiArray and decimal128ScaleArray functions
- Eliminate all AddInplace/MinusInplace from arithmetic hot paths

### CODE REVIEW

**Review by ouyuanning — APPROVED** [review]

File: pkg/sql/plan/function/arith_decimal_fast.go [hunk 1]
```diff
@@ -0,0 +1,1939 @@
...
+	xhi, yhi := x.B64_127, y.B64_127
+	// Fast path: both fit in int64 (B64_127 is 0 or ^0).
+	if (xhi+1) <= 1 && (yhi+1) <= 1 {
+		// Extract sign: bit 63 of B64_127.
+		neg := (xhi ^ yhi) >> 63
+		// Get absolute values of int64 representation.
+		xi := int64(x.B0_63)
+		yi := int64(y.B0_63)
+		ax := uint64((xi ^ (xi >> 63)) - (xi >> 63)) // branchless abs
+		ay := uint64((yi ^ (yi >> 63)) - (yi >> 63))
```
Comment (XuPeng-SH) [inline pkg/sql/plan/function/arith_decimal_fast.go]: ### 🔴 CRITICAL: Fast-path condition `(xhi+1) <= 1` is too loose — produces wrong results

**Bug**: The condition `(xhi+1) <= 1` accepts D128 values where `B64_127 ∈ {0, 0xFFFFFFFFFFFFFFFF}`, meaning the 128-bit value is in range `[-2^64, 2^64-1]`. But then line 1034 does `xi := int64(x.B0_63)` and line 1037 computes `abs(xi)`, which only works correctly for values in **int64 range** `[-2^63, 2^63-1]`.

**Failure case**: Positive D128 value `2^63 + 1` (B0_63=0x8000000000000001, B64_127=0):
- `int64(B0_63)` = `-(2^63 - 1)` ← WRONG, value is positive
- `abs(-(2^63-1))` = `2^63 - 1` ← magnitude off by 2
- **Result: `(2^63+1) × 1 = 2^63 - 1`** ← WRONG

**Verified** with standalone Go program:
```go
x := types.Decimal128{B0_63: 0x8000000000000001, B64_127: 0} // = 2^63+1
y := types.Decimal128{B0_63: 1, B64_127: 0} // = 1
// d128MulInline returns {B0_63: 0x7FFFFFFFFFFFFFFF, B64_127: 0} = 2^63-1  ← OFF BY 2
```

**Reachability**: `DECIMAL(19,0)` holds up to ~10^19 ≈ 1.08×2^63, so values slightly above 2^63 are valid. Also reachable from `DECIMAL(20+,0)`.

**Boundary**: Value `2^63` (B0_63=0x8000000000000000) happens to work because `abs(INT64_MIN)` wraps to `2^63` on two's complement. But `2^63+1` through `2^64-1` all fail.

**Fix**: Tighten the fast-path to only accept values that truly fit in int64 — the sign extension of B0_63 must reproduce B64_127:
```go
xFits := uint64(int64(x.B0_63) >> 63) == x.B64_127
yFits := uint64(int64(y.B0_63) >> 63) == y.B64_127
if xFits && yFits {
    // ... existing fast-path code is correct for values in [-2^63, 2^63-1]
}
```
This is branchless, has no extra memory ops, and correctly rejects values with |value| ∈ (2^63, 2^64).

**Review by XuPeng-SH — COMMENTED** [review]
## Comprehensive Review: Decimal Performance Optimizations

Excellent performance work — the batch-kernel architecture and branchless arithmetic are well designed, and the TPC-H Q1 speedups are substantial. The code removals are clean with no orphaned callers.

However, I found **one critical correctness bug** in `d128MulInline`, **one overflow-checking regression** in `d128Add/Sub`, and **one bug in the Distinct SUM path**. Details below.

---

### ✅ What looks good

1. **d64MulInline** — branchless D64×D64→D128 via `bits.Mul64` is correct and elegant.
2. **Coercion fix** (list_operator.go) — D64±intN staying D64 instead of promoting to D128 is well-guarded (`hasD64 && noD128 && noU64`).
3. **Division sign fix** — the test result changes (`1.796... → -1.796...`) are genuine bug fixes.
4. **D64 SUM fast path** — count-based null tracking with `Add128Unchecked` is safe (need >10^19 max-value rows to overflow a D128 accumulator).
5. **D128 SUM overflow check** — correctly conditional on `param.Width > 28` in `batchFill` and `BatchMerge`.
6. **Code removals** — `decimal128AddArray/SubArray/MultiArray/ScaleArray` and old `baseTemplate.go` functions are fully replaced. No dead callers remain.
7. **D64→D128 cast optimization** (func_cast.go) — branchless sign extension with direct slice writes is correct.

---

### Summary of Issues

| Severity | Issue | Location |
|----------|-------|----------|
| 🔴 CRITICAL | d128MulInline fast-path accepts values outside int64 range | arith_decimal_fast.go:1030 |
| 🟠 HIGH | d128Add/Sub overflow checking regression | arith_decimal_fast.go:692 |
| 🟠 HIGH | Distinct D128 SUM flush ignores overflowCheck | sum_decimal_fast.go:429 |
| 🟡 MEDIUM | Missing edge-case tests near 2^63 boundary | arith_decimal_fast_test.go |
| 🟡 MEDIUM | Pow10[] bounds not validated in scale helpers | arith_decimal_fast.go |

File: pkg/sql/plan/function/arith_decimal_fast.go [hunk 2]
```diff
@@ -0,0 +1,1939 @@
...
+	len2 := len(v2)
+	noNull := rsnull.IsEmpty()
+
+	var carry uint64
+
+	if len1 == len2 {
+		if noNull {
+			for i := 0; i < len1; i++ {
+				rs[i].B0_63, carry = bits.Add64(v1[i].B0_63, v2[i].B0_63, 0)
+				rs[i].B64_127, _ = bits.Add64(v1[i].B64_127, v2[i].B64_127, carry)
```
Comment (XuPeng-SH) [inline pkg/sql/plan/function/arith_decimal_fast.go]: ### 🟠 HIGH: Overflow checking regression — old `AddInplace` detected same-sign overflow, new code doesn't

The old `Decimal128.AddInplace()` method checked for same-sign overflow via sign-change detection:
```go
// Old code (decimal.go AddInplace):
if signx == y.Sign() {
    x.B0_63, carryout = bits.Add64(x.B0_63, y.B0_63, 0)
    x.B64_127, _ = bits.Add64(x.B64_127, y.B64_127, carryout)
    if signx != x.Sign() {  // ← overflow check
        err = moerr.NewInvalidInputNoCtxf("Decimal128 Add overflow: ...")
    }
}
```

The new `d128AddSameScale` uses raw `bits.Add64` without any overflow detection. This means `SELECT a + b` where both `a` and `b` are near-max `DECIMAL(38,0)` values will now **silently produce a wrong result** instead of returning an error.

I understand this is a deliberate performance trade-off, but it's a behavior regression. Suggestion: either:
1. Add the sign-change check back (one comparison + one branch per element, minimal perf impact), or
2. Document this as a known limitation in the function comment so future readers understand it's intentional.

File: pkg/sql/colexec/aggexec/sum_decimal_fast.go [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+					if err := vector.AppendNull(vecs[i], exec.mp); err != nil {
+						return nil, err
+					}
+					continue
+				}
+				var sum types.Decimal128
+				xcnt := 0
+				err := exec.state[i].iter(uint16(j), func(k []byte) error {
+					ptr := util.UnsafeFromBytes[types.Decimal128](k[kAggArgPrefixSz:])
+					sum = sum.Add128Unchecked(*ptr)
```
Comment (XuPeng-SH) [inline pkg/sql/colexec/aggexec/sum_decimal_fast.go]: ### 🟠 HIGH: Distinct D128 SUM flush ignores `overflowCheck` — uses `Add128Unchecked` unconditionally

The `batchFill` (line 309) and `BatchMerge` (line 379) paths correctly condition on `exec.overflowCheck` to choose between `Add128` (checked) and `Add128Unchecked`. But the Distinct flush path here always uses `Add128Unchecked`, even when `overflowCheck=true` (precision > 28).

This means `SELECT SUM(DISTINCT col) FROM t` on a `DECIMAL(38,0)` column can silently overflow.

**Fix**: respect the flag:
```go
if exec.overflowCheck {
    sum, err = sum.Add128(*ptr)
    if err != nil { return nil, err }
} else {
    sum = sum.Add128Unchecked(*ptr)
}
```

File: pkg/sql/plan/function/arith_decimal_fast.go [hunk 4]
```diff
@@ -0,0 +1,1939 @@
...
+		return d128AddSameScale(v1, v2, rs, rsnull)
+	}
+	return d128AddDiffScale(v1, v2, rs, scale1, scale2, rsnull)
+}
+
+// d128Sub is the batch kernel for Decimal128 subtraction.
+// Same-scale: direct bits.Sub64 without overflow checking.
+// Different-scale: scales operand(s) first, then uses bits.Sub64 loop.
+
+func d128AddSameScale(v1, v2, rs []types.Decimal128, rsnull *nulls.Nulls) error {
```
Comment (XuPeng-SH) [inline pkg/sql/plan/function/arith_decimal_fast.go]: ### 🟡 MEDIUM: Same overflow concern applies to `d128SubSameScale`

Just noting the same overflow-checking regression applies to the subtraction path as well (`a - (-b)` where both have the same sign after negation). The old `SubInplace` had the same sign-change detection.

File: pkg/sql/plan/function/arith_decimal_fast.go [hunk 5]
```diff
@@ -0,0 +1,1939 @@
...
+					return err
+				}
+			}
+		}
+	} else if len1 == 1 {
+		a := v1[0]
+		if rsnull.IsEmpty() {
+			for i := 0; i < len2; i++ {
+				var err error
+				rs[i], err = a.Add64(v2[i])
```
Comment (XuPeng-SH) [inline pkg/sql/plan/function/arith_decimal_fast.go]: ### 🟡 MEDIUM: `Pow10` array bounds

The `Pow10` table has 20 entries (indices 0–19). The scale helpers like `d128ScaleUp` use `Pow10[scaleDiff]` without bounds validation. While `scaleDiff` should always be in range due to SQL type rules, a defensive bounds check (or at least a comment documenting the invariant) would prevent cryptic panics if the invariant is ever violated upstream.

**Review by heni02 — APPROVED** [review]

**Comment by aunjgr** [comment]
**Re: list_operator.go:1804 coercion concern**: No change needed — int64 fits in Decimal64 (same 64-bit storage, scale 0). The `noU64` check covers the only risky type (`uint64 > INT64_MAX`). int64→Decimal64 is a no-op reinterpret.

**Fixed in ed1e9d69a:**
- d128DivInline: `bits.Add64` carry check for `zHi += crossLo` (prevents silent scaling overflow)
- Removed dead `xcnt` variable from D64/D128 DISTINCT SUM Flush paths

File: pkg/sql/colexec/aggexec/sum_decimal_fast.go [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+		return exec.batchFillArgs(offset, groups, vectors, true)
+	}
+	return exec.batchFill(offset, groups, vectors)
+}
+
+// batchFill is the hot-path for both SUM(decimal64) and AVG(decimal64).
+// Uses count tracking instead of null bitmap checks on accumulator.
+func (exec *sumDecimal64FastExec) batchFill(offset int, groups []uint64, vectors []*vector.Vector) error {
+	vec := vectors[0]
+	vals := vector.MustFixedColNoTypeCheck[types.Decimal64](vec)
```
[the remark this answers is not in the document]
Comment (aunjgr) [reply pkg/sql/colexec/aggexec/sum_decimal_fast.go:85]: Fixed in 0f427443d — uses branchless `constMask` so const vectors always index 0: `vals[idx&constMask]`.

File: pkg/sql/colexec/aggexec/sum_decimal_fast.go [hunk 7]
```diff
@@ -0,0 +1,489 @@
...
+	if exec.IsDistinct() {
+		return exec.batchFillArgs(offset, groups, vectors, true)
+	}
+	return exec.batchFill(offset, groups, vectors)
+}
+
+func (exec *sumDecimal128FastExec) batchFill(offset int, groups []uint64, vectors []*vector.Vector) error {
+	vec := vectors[0]
+	vals := vector.MustFixedColNoTypeCheck[types.Decimal128](vec)
+	hasNull := vec.HasNull()
```
[the remark this answers is not in the document]
Comment (aunjgr) [reply pkg/sql/colexec/aggexec/sum_decimal_fast.go:308]: Fixed in same commit (0f427443d) — same `constMask` pattern applied to both overflow-checked and unchecked paths.

File: pkg/sql/colexec/aggexec/sum_decimal_fast.go [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+	overflowCheck bool // true when input precision > 28 (SUM can overflow)
+}
+
+func newSumDecimal128FastExec(mp *mpool.MPool, isSum bool, aggID int64, isDistinct bool, param types.Type) AggFuncExec {
+	var exec sumDecimal128FastExec
+	exec.mp = mp
+	exec.isSum = isSum
+	// Precision > 28: SUM of 17B max-value rows can overflow Decimal128.
+	exec.overflowCheck = param.Width > 28
+	sumTyp := SumReturnType([]types.Type{param})
```
[the remark this answers is not in the document]
Comment (aunjgr) [reply pkg/sql/colexec/aggexec/sum_decimal_fast.go:263]: Improved the comment in 0f427443d to document the safety bound: width ≤ 28 means max input ≈ 10^28, so overflow requires ~17 billion rows per group — well beyond practical limits. The checked path (`Add128`) is used for width > 28.

**Review by XuPeng-SH — APPROVED** [review]
